plutus-playground/plutus-playground-server/usecases/Game.hs

```diff
@@ -31,29 +37,87 @@ mkRedeemerScript word =
 
+-- | The validator script of the game. 
 gameValidator :: ValidatorScript
 gameValidator = ValidatorScript (Ledger.fromCompiledCode $$(PlutusTx.compile [||
+    -- The code between the '[||' and  '||]' quotes is on-chain code.
     \(ClearString guess) (HashedString actual) (p :: PendingTx') ->
 
     if $$(P.equalsByteString) actual ($$(P.sha2_256) guess)
+        -- inside the on-chain code we can write $$(P.xxx) to import functions
+        -- from the PlutusTx Prelude (imported qualified at the top of the 
+        -- module)
+
     then ()
-    else $$(P.traceH) "WRONG!" ($$(P.error) ())
+    else ($$(P.error) ($$(P.traceH) "WRONG!" ()))
 
     ||]))
 
+-- | The address of the game (the hash of its validator script)
 gameAddress :: Address'
 gameAddress = Ledger.scriptAddress gameValidator
 
+-- | The "lock" contract endpoint. See note [Contract endpoints]
 lock :: String -> Value -> MockWallet ()
-lock word value = payToScript_ gameAddress value (mkDataScript word)
-
+lock word value = 
+    payToScript_ gameAddress value (mkDataScript word)
+    -- 'payToScript_' is a function of the wallet API. It takes a script 
+    -- address, a value and a data script, and submits a transaction that
+    -- pays the value to the address, using the data script.
+    --
+    -- The underscore at the end of the name indicates that 'payToScript_'
+    -- discards its result. If you want to hold on to the transaction you can
+    -- use 'payToScript'.
+
+-- | The "guess" contract endpoint. See note [Contract endpoints]
 guess :: String -> MockWallet ()
-guess word = collectFromScript gameValidator (mkRedeemerScript word)
-
--- | Tell the wallet to start watching the address of the game script
+guess word = 
+    collectFromScript gameValidator (mkRedeemerScript word)
+    -- 'collectFromScript' is a function of the wallet API. It consumes the 
+    -- unspent transaction outputs at a script address and pays them to a 
+    -- public key address owned by this wallet. It takes the validator script 
+    -- and the redeemer scripts as arguments.
+    --
+    -- Note that before we can use 'collectFromScript', we need to tell the 
+    -- wallet to start watching the address for transaction outputs (because 
+    -- the wallet does not keep track of the UTXO set of the entire chain).
+
+-- | The "startGame" contract endpoint, telling the wallet to start watching 
+--   the address of the game script. See note [Contract endpoints]
 startGame :: MockWallet ()
-startGame = startWatching gameAddress
+startGame = 
+    startWatching gameAddress
+    -- 'startWatching' is a function of the wallet API. It instructs the wallet 
+    -- to keep track of all outputs at the address. Player 2 needs to call 
+    -- 'startGame' before Player 1 uses the 'lock' endpoint, to ensure that 
+    -- Player 2's wallet is aware of the game address.
 
 $(mkFunction 'lock)
 $(mkFunction 'guess)
 $(mkFunction 'startGame)
+
+{- Note [Contract endpoints]
```

Comment:
Not sure whether we should repeat this note in the other examples? In general, some of the docs about library functions, endpoints etc. could be in all of them, but it would be annoying to copy-paste it so I'm not sure what we  should do.

Reply:
I put them in the order that I mentioned them in the demo (Game contract first, which introduced endpoints, then crowdfunding, which introduced blockchain triggers). But this should really live in the Playground manual which doesn't exist yet.